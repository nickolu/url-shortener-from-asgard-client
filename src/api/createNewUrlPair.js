import corsFetch from "./corsFetch";
const createNewUrlPair = async ({ longUrl, userId }) => {
    const response = await corsFetch(
        `create-shortened-url?original_url=${longUrl}&user_id=${userId}`
    );

    return response.json();
};

export default createNewUrlPair;
