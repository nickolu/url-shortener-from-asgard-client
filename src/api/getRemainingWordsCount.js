import corsFetch from "./corsFetch";

const getRemainingWordsCount = async () => {
    const response = await corsFetch("get-remaining-asgardian-words-count");

    return response.json();
};

export default getRemainingWordsCount;
