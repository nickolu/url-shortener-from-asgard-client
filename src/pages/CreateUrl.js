import React, { Component } from "react";
import PropTypes from "prop-types";
import createNewUrlPair from "../api/createNewUrlPair";
import getRemainingWordsCount from "../api/getRemainingWordsCount";

class CreateUrl extends Component {
    constructor() {
        super();

        this.state = {
            longUrl: "",
            shortUrl: "",
            remainingWordsCount: null
        };

        this.setLongUrl = this.setLongUrl.bind(this);
        this.getUrlPair = this.getUrlPair.bind(this);
        this.getRemainingWordsCount = this.getRemainingWordsCount.bind(this);
    }

    setLongUrl(event) {
        const longUrl = event.target.value;

        this.setState({ longUrl: "https://" + longUrl });
    }

    async getUrlPair() {
        const longUrl = this.state.longUrl;
        const userId = this.props.userId;
        const urlPair = await createNewUrlPair({ longUrl, userId });

        if (this.state.longUrl) {
            this.setState(
                { shortUrl: urlPair.shortUrl },
                this.getRemainingWordsCount
            );
        }
    }

    async getRemainingWordsCount() {
        const remainingWordsCount = await getRemainingWordsCount();

        this.setState({ remainingWordsCount: remainingWordsCount.remaining });
    }

    componentDidMount() {
        this.getRemainingWordsCount();
    }

    render() {
        return (
            <div>
                <h2>Create a Shortened URL</h2>
                {this.state.remainingWordsCount !== null && (
                    <p>
                        {this.state.remainingWordsCount} unique Asgardian words
                        remain!
                    </p>
                )}
                {this.props.userId ? (
                    <div>
                        <strong>Your Url:</strong> https://
                        <input onChange={this.setLongUrl} />
                        <button onClick={this.getUrlPair}>Submit</button>
                    </div>
                ) : (
                    <div>Loading User Id</div>
                )}

                {this.state.shortUrl && (
                    <div data-jest-selector="short-url-result">
                        Here's your short url!:
                        <ul>
                            <li>Your Url: {this.state.longUrl}</li>
                            <li>
                                Shortened Url:{" "}
                                <a
                                    href={
                                        "http://127.0.0.1:5000/" +
                                        this.state.shortUrl
                                    }
                                >
                                    {"http://127.0.0.1:5000/" +
                                        this.state.shortUrl}
                                </a>
                            </li>
                        </ul>
                    </div>
                )}
            </div>
        );
    }
}

CreateUrl.propTypes = {
    userId: PropTypes.string
};

export { CreateUrl };
