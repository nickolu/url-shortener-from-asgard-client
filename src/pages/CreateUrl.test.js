jest.mock("../api/createNewUrlPair");
jest.mock("../api/getRemainingWordsCount");
import React from "react";
import { mount } from "enzyme";
import { CreateUrl } from "./CreateUrl";
import createNewUrlPair from "../api/createNewUrlPair";
import getRemainingWordsCount from "../api/getRemainingWordsCount";
import wait from "waait";

createNewUrlPair.mockImplementation(({ longUrl }) => ({
    shortUrl: "helloWorld",
    longUrl
}));

getRemainingWordsCount.mockImplementation(() => ({ remaining: 100 }));

const FAKE_USER_ID = "1";

let wrapper;
describe("CreateUrl", () => {
    beforeEach(() => {
        wrapper = mount(<CreateUrl userId={FAKE_USER_ID} />);
    });

    it("should render", () => {
        expect(wrapper.render()).toMatchSnapshot();
    });

    it("should show the result after submitting a url", async () => {
        const event = { target: { value: "www.google.com" } };

        wrapper.find("input").simulate("change", event);
        wrapper.find("button").simulate("click", event);
        await wait();
        wrapper.update();

        expect(createNewUrlPair).toHaveBeenCalledTimes(1);
        expect(
            wrapper.find('[data-jest-selector="short-url-result"]').text()
        ).toContain("Here's your short url!");

        expect(
            wrapper.find('[data-jest-selector="short-url-result"]').text()
        ).toContain("Here's your short url!");
    });
});
