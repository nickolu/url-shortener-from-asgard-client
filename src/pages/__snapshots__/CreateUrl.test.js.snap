// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateUrl should render 1`] = `
<div>
  <h2>
    Create a Shortened URL
  </h2>
  <p>
    100 unique Asgardian words remain!
  </p>
  <div>
    <strong>
      Your Url:
    </strong>
     https://
    <input />
    <button>
      Submit
    </button>
  </div>
</div>
`;
